pkg/api/helpers.go

```diff
@@ -0,0 +1,46 @@
+package api
+
+import "time"
+
+// RecordStageAndStepInfo records details about each build stage and step
+func RecordStageAndStepInfo(stages Stages, stageName StageName, stepName StepName, startTime time.Time, endTime time.Time) Stages {
+	// Make sure that the stages slice is initialized
+	if len(stages) == 0 {
+		stages = make([]StageInfo, 0)
+	}
+
+	// If the stage already exists  update the endTime and Duration, and append the new step.
+	for stageKey, stageVal := range stages {
+		if stageVal.StageName == stageName {
+			stages[stageKey].EndTime = endTime
+			stages[stageKey].Duration = stages[stageKey].Duration + endTime.Sub(startTime)
```

Comment:
just record stage duration as:  startTime of the (endtime of the last step)-(starttime of the first step).

(and get rid of the endtime field, per my comment below)